traversal/GraphTraversal.java

```diff
@@ -178,39 +178,57 @@ public void filter(Set<? extends Identifier.Variable.Retrievable> filter) {
 
     public static class Thing extends GraphTraversal {
 
-        private List<Planner> planners;
+        private Map<Structure, Planner> planners;
         private boolean modifiable;
 
         public Thing() {
             super();
             modifiable = true;
+            planners = new HashMap<>();
+        }
+
+        private void initialise(PlannerCache cache) {
+            if (planners.isEmpty()) {
+                structures().forEachRemaining(structure -> {
+                    Planner planner = cache.get(structure, Planner::create);
+                    planners.put(structure, planner);
+                });
+            }
```

Comment:
in order to update the cache we  now have to remember which structure is associated with which planner